triplea-game/triplea
# Add mp3 support

Change Summary:
- Add a test for ClipPlayer
- Refactor ClipPlayer, simplify it
- Performance fix when muting or unmuting all sounds (3aafc0b)
- Add mp3zoom libraries and update code to be able to play mp3 files


Review note:
- Much of the changes are simply to clean up the code so that it can be updated reasonably
  - bbd55b3b9e97e5294990e4cf92f6711c6290e907 does the actual mp3 support
  - Cleaned up code is mainly public API, or interfaces that have unused methods or excessive and unused parameters
  - The code to test if a sound could be played and to play a sound was copy pasted, one clean up was to combine this functionality so only one code path actually then plays sounds

To test the performance fix:
- start a game, go to sound options
- use the select all or select none check boxes, click okay. Notice that the UI will freeze up for a few seconds (FYI, with a fast HD you may not notice this delay - after upgrading to a SSD disk that is a full 1000x faster than the Sata drive it replaced, I no longer noticed this delay)

To test mp3 file support:
- Make sure you can hear game sounds first (Napoleonic start sound is my favorite)
- Check out this branch:
  - `git remote add DanV https://github.com/DanVanAtta/triplea.git`
  - `git fetch DanV`
  - `git checkout clipPlayer`
- Add the three new jar libraries to your eclipse path (mp3*.jar, jl1*.jar, tritonus*.jar)
- Repeat the first experiment, this confirms wav files still play
- Use your favorite mp3 encoding utility to convert the test sound to mp3. `lame` is the tool I used, the napolean start sound is under "assets/preindustrial/game_start". Audacity is another tool I think available for windows that could do the same thing with a GUI (it comes with a lame plugin)

Two gotchyas for mp3 file testing:
- Adding the jar libraries is needed at runtime, not compile time. So you won't get any compile errors if they are not on your path or any other kind of reminder to do that.
- If sound is muted in the game then you won't hear anything, both the sound option needs to be unmuted and "sounds enabled" checked! A few times after coming back from playing I ran into that the hard way.


[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/201)
